File: acmOtherMaterials/Ulm Local Contest/2002/solution/allinall.hs
-- Problem   All in All
-- Algorithm Greedy
-- Runtime   O(n)
-- Author    Walter Guttmann
-- Date      26.12.2001

main :: IO ()
main =
  do input <- readFile "allinall.in"
     mapM_ solve $ groupn 2 $ words input

groupn :: Int -> [a] -> [[a]]
groupn n [] = []
groupn n xs = let (ys,zs) = splitAt n xs in ys : groupn n zs

solve :: [String] -> IO ()
solve [x,y] = putStrLn $ if isSubSeq x y then "Yes" else "No"

isSubSeq :: Eq a => [a] -> [a] -> Bool
isSubSeq [] _ = True
isSubSeq _ [] = False
isSubSeq (x:xs) (y:ys) = if x == y then isSubSeq xs ys else isSubSeq (x:xs) ys

File: acmOtherMaterials/Ulm Local Contest/2002/solution/allinall.java
// Problem   All in All
// Algorithm Greedy
// Runtime   O(n)
// Author    Walter Guttmann
// Date      26.12.2001

import java.io.*;

public class allinall {
  public static void main (String[] arg) throws Exception
  {
    BufferedReader in = new BufferedReader (new FileReader ("allinall.in"));
    StreamTokenizer st = new StreamTokenizer (in);
    st.ordinaryChars ('-', '9');
    st.wordChars ('-', '9');
    while (st.nextToken () != st.TT_EOF)
    {
      String s = st.sval;
      st.nextToken();
      String t = st.sval;
      int is = 0;
      for (int it = 0 ; it < t.length() ; it++)
        if (is < s.length() && s.charAt (is) == t.charAt (it))
          is++;
      System.out.println (is == s.length() ? "Yes" : "No");
    }
  }
}

